src/servermodule.h

```diff
@@ -0,0 +1,1695 @@
+#define SERVERMODULE_REPLY_NULL 4
+#define SERVERMODULE_REPLY_MAP 5
+#define SERVERMODULE_REPLY_SET 6
+#define SERVERMODULE_REPLY_BOOL 7
+#define SERVERMODULE_REPLY_DOUBLE 8
+#define SERVERMODULE_REPLY_BIG_NUMBER 9
+#define SERVERMODULE_REPLY_VERBATIM_STRING 10
+#define SERVERMODULE_REPLY_ATTRIBUTE 11
+#define SERVERMODULE_REPLY_PROMISE 12
+
+/* Postponed array length. */
+#define SERVERMODULE_POSTPONED_ARRAY_LEN -1  /* Deprecated, please use SERVERMODULE_POSTPONED_LEN */
+#define SERVERMODULE_POSTPONED_LEN -1
+
+/* Expire */
+#define SERVERMODULE_NO_EXPIRE -1
+
+/* Sorted set API flags. */
+#define SERVERMODULE_ZADD_XX      (1<<0)
+#define SERVERMODULE_ZADD_NX      (1<<1)
+#define SERVERMODULE_ZADD_ADDED   (1<<2)
+#define SERVERMODULE_ZADD_UPDATED (1<<3)
+#define SERVERMODULE_ZADD_NOP     (1<<4)
+#define SERVERMODULE_ZADD_GT      (1<<5)
+#define SERVERMODULE_ZADD_LT      (1<<6)
+
+/* Hash API flags. */
+#define SERVERMODULE_HASH_NONE       0
+#define SERVERMODULE_HASH_NX         (1<<0)
+#define SERVERMODULE_HASH_XX         (1<<1)
+#define SERVERMODULE_HASH_CFIELDS    (1<<2)
+#define SERVERMODULE_HASH_EXISTS     (1<<3)
+#define SERVERMODULE_HASH_COUNT_ALL  (1<<4)
+
+#define SERVERMODULE_CONFIG_DEFAULT 0 /* This is the default for a module config. */
+#define SERVERMODULE_CONFIG_IMMUTABLE (1ULL<<0) /* Can this value only be set at startup? */
+#define SERVERMODULE_CONFIG_SENSITIVE (1ULL<<1) /* Does this value contain sensitive information */
+#define SERVERMODULE_CONFIG_HIDDEN (1ULL<<4) /* This config is hidden in `config get <pattern>` (used for tests/debugging) */
+#define SERVERMODULE_CONFIG_PROTECTED (1ULL<<5) /* Becomes immutable if enable-protected-configs is enabled. */
+#define SERVERMODULE_CONFIG_DENY_LOADING (1ULL<<6) /* This config is forbidden during loading. */
+
+#define SERVERMODULE_CONFIG_MEMORY (1ULL<<7) /* Indicates if this value can be set as a memory value */
+#define SERVERMODULE_CONFIG_BITFLAGS (1ULL<<8) /* Indicates if this value can be set as a multiple enum values */
+
+/* StreamID type. */
+typedef struct RedisModuleStreamID {
+    uint64_t ms;
+    uint64_t seq;
+} RedisModuleStreamID;
+
+/* StreamAdd() flags. */
+#define SERVERMODULE_STREAM_ADD_AUTOID (1<<0)
+/* StreamIteratorStart() flags. */
+#define SERVERMODULE_STREAM_ITERATOR_EXCLUSIVE (1<<0)
+#define SERVERMODULE_STREAM_ITERATOR_REVERSE (1<<1)
+/* StreamIteratorTrim*() flags. */
+#define SERVERMODULE_STREAM_TRIM_APPROX (1<<0)
+
+/* Context Flags: Info about the current context returned by
+ * RM_GetContextFlags(). */
+
+/* The command is running in the context of a Lua script */
+#define SERVERMODULE_CTX_FLAGS_LUA (1<<0)
+/* The command is running inside a Redis transaction */
+#define SERVERMODULE_CTX_FLAGS_MULTI (1<<1)
+/* The instance is a master */
+#define SERVERMODULE_CTX_FLAGS_MASTER (1<<2)
+/* The instance is a slave */
+#define SERVERMODULE_CTX_FLAGS_SLAVE (1<<3)
+/* The instance is read-only (usually meaning it's a slave as well) */
```

Comment:
My one maybe bold take, we should update this:
```suggestion
/* The instance is a replica */
#define SERVERMODULE_CTX_FLAGS_REPLICA (1<<3)
/* The instance is read-only (usually meaning it's a replica as well) */
```

We said we don't use slave except for compatibility, and this isn't strictly compatibility.

Reply:
Change MASTER flags too then? Or shall we leave that for later/never?

Reply:
If we do change to REPLICA for this flag, then the docs for migrating from redismodule to valkeymodule needs to mention this explicitly, apart from the s/REDISMODULE/VALKEYMODULE/g and s/RedisModule/ValkeyModule/g.

@PingXie Can you add this as a comment near the top of valkeymodule.h?

Reply:
I thought about MASTER, but we didn't pick a new term. If we can align on that new term I would like to change it to.

Reply:
>I thought about MASTER, but we didn't pick a new term. If we can align on that new term I would like to change it to.

I picked "primary" as it seems to have been established in both the code base and the documentation. 

>@PingXie Can you add this "migration guide" as a comment near the top of valkeymodule.h?

Let me add this in the next commit

Reply:
done